pydantic/_internal/_core_utils.py

```diff
@@ -271,115 +271,220 @@ def handle_definitions_schema(self, schema: core_schema.DefinitionsSchema) -> Co
         return schema
 
-    def handle_arguments_schema(self, schema: core_schema.ArgumentsSchema) -> CoreSchema:
-        replaced_arguments_schema = []
+    def handle_arguments_schema(self, schema: core_schema.ArgumentsSchema, f: Walk) -> core_schema.CoreSchema:
+        replaced_arguments_schema: list[core_schema.ArgumentsParameter] = []
         for param in schema['arguments_schema']:
             replaced_param = param.copy()
-            replaced_param['schema'] = self._walk(param['schema'])
+            replaced_param['schema'] = self.walk(param['schema'], f)
             replaced_arguments_schema.append(replaced_param)
         schema['arguments_schema'] = replaced_arguments_schema
         if 'var_args_schema' in schema:
-            schema['var_args_schema'] = self._walk(schema['var_args_schema'])
+            schema['var_args_schema'] = self.walk(schema['var_args_schema'], f)
         if 'var_kwargs_schema' in schema:
-            schema['var_kwargs_schema'] = self._walk(schema['var_kwargs_schema'])
+            schema['var_kwargs_schema'] = self.walk(schema['var_kwargs_schema'], f)
         return schema
 
-    def handle_call_schema(self, schema: core_schema.CallSchema) -> CoreSchema:
-        schema['arguments_schema'] = self._walk(schema['arguments_schema'])
+    def handle_call_schema(self, schema: core_schema.CallSchema, f: Walk) -> core_schema.CoreSchema:
+        schema['arguments_schema'] = self.walk(schema['arguments_schema'], f)
         if 'return_schema' in schema:
-            schema['return_schema'] = self._walk(schema['return_schema'])
+            schema['return_schema'] = self.walk(schema['return_schema'], f)
         return schema
+
+
+_dispatch = _WalkCoreSchema().walk
+
+
+def walk_core_schema(schema: core_schema.CoreSchema, f: Walk) -> core_schema.CoreSchema:
+    """Recursively traverse a CoreSchema.
+
+    Args:
+        schema (core_schema.CoreSchema): The CoreSchema to process, it will not be modified.
+        f (Walk): A function to apply. This function takes two arguments:
+          1. The current CoreSchema that is being processed
+             (not the same one you passed into this function, one level down).
+          2. The "next" `f` to call. This lets you for example use `f=functools.partial(some_method, some_context)`
+             to pass data down the recursive calls without using globals or other mutable state.
+
+    Returns:
+        core_schema.CoreSchema: A processed CoreSchema.
+    """
+    return f(schema.copy(), _dispatch)
+
+
+def simplify_schema_references(schema: core_schema.CoreSchema) -> core_schema.CoreSchema:  # noqa: C901
+    """
+    Simplify schema references by:
+      1. Grouping all definitions into a single top-level `definitions` schema, similar to a JSON schema's `#/$defs`.
+      2. Inlining any definitions that are only referenced in one place and are not involved in a cycle.
+      3. Removing any unused `ref` references from schemas.
+    """
```

Comment:
Once we update pydantic-core I'll make another PR to replace all of the above functions, and the ref stuff in json schema and tagged unions, with this thing.